napari/util/misc.py

```diff
@@ -351,3 +353,80 @@ def __str__(self):
 
 
 callsignature = CallSignature.from_callable
+
+
+class ReadOnlyWrapper(wrapt.ObjectProxy):
+    """
+    Disable item and attribute setting with the exception of  ``__wrapped__``.
+    """
+
+    def __setattr__(self, name, val):
+        if name != '__wrapped__':
+            raise TypeError(f'cannot set attribute {name}')
+        super().__setattr__(name, val)
+
+    def __setitem__(self, name, val):
+        raise TypeError(f'cannot set item {name}')
+
+
+def mouse_press_callbacks(obj, event):
+    """Run mouse press callbacks on either layer or viewer object.
+
+    Parameters
+    ---------
+    obj : napari.components.ViewerModel or napar.layers.Layer
+        Layer or Viewer object to run callbacks on
+    event : Event
+        Mouse event
+    """
+    for mouse_drag_func in obj.mouse_drag_callbacks:
+        gen = mouse_drag_func(obj, event)
+        if inspect.isgeneratorfunction(mouse_drag_func):
+            try:
+                next(gen)
+                obj._mouse_drag_gen[mouse_drag_func] = gen
+                obj._persisted_mouse_event[gen] = event
+            except StopIteration:
+                pass
```

Comment:
So imho all this logic needs comments explaining what is being done, though I sort of get it:

- run the function, getting a result
- if the result is a generator, then get the next value from the generator, and
- ??? What is being added to those dictionaries? And where are the dictionaries used?

In general, one of the principles of writing bug free code is that code should be as local as possible, meaning that I shouldn't have to keep in my head all the possible accesses and lifetimes of stuff that lives outside of this function. I'm not sure how to improve this implementation but at a minimum, there should be comments here about what those dictionaries do and where they are "cleared" and why. (I see some dels in the next function but it's unclear to me whether they refer to the same object (ie are they supposed to happen in quick succession)

Reply:
done